cmd/crank/describe.go

```diff
@@ -0,0 +1,96 @@
+
+import (
+	"fmt"
+	"os"
+
+	"golang.org/x/exp/slices"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	"github.com/crossplane/crossplane/cmd/crank/printer"
+	"github.com/crossplane/crossplane/internal/k8s"
+)
+
+const (
+	errGetResource = "Couldn't get requested resource."
+	errCliOutput   = "Error printing CLI table."
+	errGraphOutput = "Error generating graph PNG."
+)
+
+// describeCmd describes a Kubernetes Crossplane resource.
+type describeCmd struct {
+	Kind       string   `arg:"" required:"" help:"Kind of resource to describe."`
+	Name       string   `arg:"" required:"" help:"Name of specified resource to describe."`
+	Namespace  string   `short:"n" name:"namespace" help:"Namespace of resource to describe." default:"default"`
+	Output     string   `short:"o" name:"output" help:"Output type of graph. Possible output types: tree, table, graph." enum:"tree,table,graph" default:"tree"`
+	Fields     []string `short:"f" name:"fields" help:"Fields that are printed out in the header." default:"kind,name"`
+	OutputPath string   `short:"p" name:"path" help:"Output path for graph PNG. Only valid when set with output=graph." default:"graph.png"`
+}
+
+func (c *describeCmd) Run(logger logging.Logger) error {
+	logger = logger.WithValues("Kind", c.Kind, "Name", c.Name)
+
+	allowedFields := []string{"parent", "name", "kind", "namespace", "apiversion", "synced", "ready", "message", "event"}
+	allowedOutput := []string{"tree", "table", "graph"}
+
+	// Check if fields are valid
+	for _, field := range c.Fields {
+		if !slices.Contains(allowedFields, field) {
+			logger.Debug("Invalid field set", "invalidField", field)
+			return fmt.Errorf("Invalid field set: %s\nField has to be one of: %s", field, allowedFields)
+		}
+	}
+
+	// Check if output format is valid
+	if !slices.Contains(allowedOutput, c.Output) {
+		logger.Debug("Invalid output set", "invalidOutput", c.Output)
+		return fmt.Errorf("Invalid output set: %s\nOutput has to be one of: %s", c.Output, allowedOutput)
+	}
+
+	// set kubeconfig
+	kubeconfig, err := ctrl.GetConfig()
+	if err != nil {
+		logger.Debug(errKubeConfig, "error", err)
+		return errors.Wrap(err, errKubeConfig)
+	}
+	logger.Debug("Found kubeconfig")
+
+	// Get client for k8s package
+	client, err := k8s.NewClient(kubeconfig)
+	if err != nil {
+		return errors.Wrap(err, "Couldn't init kubeclient")
+	}
+
+	// Get Resource object. Contains k8s resource and all its children, also as Resource.
+	root, err := k8s.GetResource(c.Kind, c.Name, c.Namespace, client)
+	if err != nil {
+		logger.Debug(errGetResource, "error", err)
+		return errors.Wrap(err, errGetResource)
+	}
+
+	// Configure printer
+	var p printer.Printer
+
+	switch c.Output {
+	case "tree":
+		p = &printer.TreePrinter{
+			Indent: "",
+			IsLast: true,
+		}
+	case "table":
+		p = &printer.TablePrinter{}
+	case "graph":
+		p = &printer.GraphPrinter{}
+	}
```

Comment:
I'd probably make "tree", "table" and "graph" constants with a dedicated type and move this switch to the printer package, doing also there the validation of the known types, something like:
```golang
func New(type string) (Printer, error) {
  switch Type(type) {
    case TypeTree:
      return &printer.TreePrinter{
			Indent: "",
			IsLast: true,
		}, nil
... etc ...
  default:
    return nil, errors.Errorf("unknown printer output type: %s", type)
  }
  ...
}
```

Reply:
I implemented your suggestion in this commit https://github.com/crossplane/crossplane/pull/4751/commits/30b979234e2b9b2dcacab017f82784ec4d25d952. Let me know if it is implemented as expected.

Reply:
And small fix https://github.com/crossplane/crossplane/pull/4751/commits/bf575090521f5d82f7829d355fb6f08394fbb547

Reply:
Yes, that's exactly what I meant! Looks good 🙏